### ISSUE TITLE
WISH: sandboxing and/or stashing of compilations when switching git branches

### ISSUE THREAD

**Issue by wolftune** [commercialhaskell/stack#1634]
Okay, this may be an unrealistic dream but could also revolutionize work for Haskell overall (being a compiled language)…

When I change back and forth between two git branches that may be distinct enough to require a large recompile for the project, I wish Stack would keep the compiled stuff around from each branch and use it accordingly. Thus, each branch would only need to compile the updates that are since the last compile of that branch rather than run a large recompile every time I switch.

Ideally, this would recognize that some files are totally unchanged and have unchanged dependencies among branches and somehow know to use that same compilation even when distinct files are sandboxed for each branch. That may not be feasible even dream sense, but it would be the ideal.

Perhaps as a practical matter, a "save this branch compilation" function could be added so that even if the whole wish idea above is impossible or unrealistic to automate, I could decide myself that I want to effectively *stash* a current compilation of the project, move to another branch, do whatever there, including recompiling, and then be able to go back to the first branch and *stash pop* so-to-speak to get back the stashed compilation and not have to recompile.

Ideally, I could stash with a name, like using memory registers to have a few stashed compilations, but a single stash at first would still be a useful step.

In the end, some sort of this working could end a lot of excessive recompiling for all Haskell development…

**Comment by mgsloan** [comment]
This is actually a quite realistic dream :) .  I was thinking about the same thing a couple days ago, infact, in terms of per-branch compilation results.

I'm temporarily dubbing this a "build cache", dunno if that name should stick.

One tricky thing is that we can't actually move `dist` dirs, and then use them in a different location, because the build depends on the absolute path.  Probably not the only thing, but `Paths_*` module provide absolute paths.  This isn't a big problem, just inconvenient.  It means that when we copy a dist dir (forking the build cache), the only way we can use it is by moving it back into the spot it came from.  This works out fine for the usecase you describe, but I'd also like to use multiple build caches at once.  This'd allow you to simultaneously build the project with different settings.  Unfortunately, this conflicts with being able to copy the build cache.

I've considered a bit how to add support for this, and I think the best approach would be to have it be a natural consequence of supporting multiple build configurations.  There isn't an issue for that yet, but there's [this comment](https://github.com/commercialhaskell/stack/issues/1568#issuecomment-169861597).  Reusing this mechanism means that we only have to specify one thing (which config), instead of two (which config + which cache).  I think this choice also implies that we'd forego being able to fork build caches (copying), in favor of being able to do simultaneous builds.

Glad to have even more reasons to have multiple configurations, as I'm starting to get rather keen on that feature, for a variety of reasons.

**Comment by 3noch** [comment]
I know talked about this long ago when were discussing custom snapshots. If custom snapshots are merged with extra-deps, I think something rather elegant could result which supports this.

**Comment by mgsloan** [comment]
Nowadays we have configurable work dirs via the `work-dir` config option / `--work-dir` cli flag.  It hink the most elegant way to do this would be to implement https://github.com/commercialhaskell/stack/issues/732 

Then, add some stack commands for updating the `work-dir` config option in the local config.  See https://github.com/commercialhaskell/stack/issues/115 for a general discussion of refactoring yaml.  For the purposes of updating this local config, i think just a parse + reserialize is fine.

Then, you could have git hooks which simply update the `stack.local.yaml` with a different `work-dir:` value.

**Comment by radix** [comment]
@mgsloan that sounds like a good idea. it would still be nicer to have something a bit more automatic, but it would definitely save me a lot of rebuilding :)

**Comment by mgsloan** [comment]
@radix True.  What I like about this is that it's flexible.  Something more automatic can be built atop

**Comment by radix** [comment]
@mgsloan one thing about your proposed workaround is that I'm not sure how to make it handle new branches -- if you create a new branch, and you're using branch-specific work-dirs, then it has to build everything from scratch.

I would really prefer to see a system that just keeps artifacts named after the hash of the source code of the package + all dependencies, a la nix, with a garbage-collection mechanism that deletes the oldest artifacts once the size of the work-dir grows beyond some locally configurable limit. I'm not sure what difficulties this engenders, but I've implemented this in other build systems and it wasn't completely intractible :) as long as there's some relatively easy way to get a list of all artifacts that go into a build.

**Comment by mgsloan** [comment]
Yes, a build system like that would be ideal.  More straightforward to distribute among multiple computers too.
